File: src/ParticleEngine/ParticleEngine/Scripts/Particle/Physics/Materials/Grass.cs
using static ParticleEngine.Particle.ParticleManager;

namespace ParticleEngine.Particle.Physics.Materials
{
    public static class Grass
    {
        public static void Calculate(int row, int column, int rowSize, int columnSize, ref ParticleManager particleWorld)
        {
            ParticleInstance self = particleWorld.particles[row, column];

            int growFlowerChance = particleWorld.GenerateRandom(0, 10000000);

            if (row - 1 > 0)
            {
                if (particleWorld.particles[row - 1, column].materialType == MaterialType.Liquid &&
                    particleWorld.particles[row, column].wetValue <= 1.0f)
                {
                    particleWorld.particles[row, column].wetValue += 0.1f;
                }

                if (particleWorld.particles[row - 1, column].material == Material.Air)
                {
                    Material flowerToGrow = Material.RedFlower;
                    bool growFlower = false;
                    if (growFlowerChance < 10)
                    {
                        flowerToGrow = Material.BlueFlower;
                        growFlower = true;
                    }
                    else if (growFlowerChance < 50)
                    {
                        flowerToGrow = Material.YellowFlower;
                        growFlower = true;
                    }
                    else if (growFlowerChance < 75)
                    {
                        flowerToGrow = Material.PinkFlower;
                        growFlower = true;
                    }
                    else if (growFlowerChance < 100)
                    {
                        growFlower = true;
                    }

                    if (growFlower)
                    {
                        particleWorld.particles[row - 1, column].material = flowerToGrow;
                    }
                }
            }

            if (column - 1 > 0)
            {
                if (particleWorld.particles[row, column - 1].materialType == MaterialType.Liquid &&
                    particleWorld.particles[row, column].wetValue <= 0.5f)
                {
                    particleWorld.particles[row, column].wetValue += 0.05f;
                }
            }

            if (column < columnSize - 1)
            {
                if (particleWorld.particles[row, column + 1].materialType == MaterialType.Liquid &&
                    particleWorld.particles[row, column].wetValue <= 0.5f)
                {
                    particleWorld.particles[row, column].wetValue += 0.05f;
                }
            }

            if (row < rowSize - 1)
            {
                if (
                    (particleWorld.particles[row + 1, column].materialType == MaterialType.Gas ||
                    particleWorld.particles[row + 1, column].material == Material.Water) &&
                    particleWorld.particles[row, column].material == self.material)
                {
                    particleWorld.particles[row + 1, column] = self;
                    particleWorld.Clear(row, column);
                }
            }

            if (column > 0 && row < rowSize - 1)
            {
                if (
                    (particleWorld.particles[row + 1, column - 1].materialType == MaterialType.Liquid ||
                    particleWorld.particles[row + 1, column - 1].materialType == MaterialType.Gas) &&
                    particleWorld.particles[row, column - 1].material == Material.Air &&
                    particleWorld.particles[row, column].material == self.material)
                {

                    particleWorld.particles[row + 1, column - 1] = self;
                    particleWorld.Clear(row, column);
                }
            }

            if (column < columnSize - 1 && row < rowSize - 1)
            {
                if (
                    (particleWorld.particles[row + 1, column + 1].materialType == MaterialType.Liquid ||
                    particleWorld.particles[row + 1, column + 1].materialType == MaterialType.Gas) &&
                    particleWorld.particles[row, column + 1].material == Material.Air &&
                    particleWorld.particles[row, column].material == self.material)
                {
                    particleWorld.particles[row + 1, column + 1] = self;
                    particleWorld.Clear(row, column);
                }
            }
        }
    }
}

File: src/ParticleEngine/ParticleEngine/include/particle/physics/materials/grass.cpp
#include <iostream>

#include "particle/particle_world.h"
#include "particle/particle_physics.h"
#include "tools/tools.h"

void calculate_grass(int row, int col, ParticleWorld* particleWorld)
{
	// define self, this value is used whenever a pixel moves, it acts as a copy of all the settings for the current pixel
	ParticleWorld::ParticleInstance self = particleWorld->getParticle(row, col);

	// generate random int for growing chance of floweres
	std::uniform_int_distribution<int> dist(0, 10000000);
	int growFlowerChance = dist(particleWorld->gen);

	if (particleWorld->canUp(row))
	{
		// increment wetness if a liquid is present (up)
		if (particleWorld->getParticle(row - 1, col).materialType == ParticleWorld::MaterialType::Liquid &&
			particleWorld->getParticle(row, col).wetnessMultiplier <= 1.0f)
		{
			particleWorld->particles[get1DIndex(row, col, particleWorld->colSize)].wetnessMultiplier += 0.1f;
		}

		if (particleWorld->getParticle(row - 1, col).material == ParticleWorld::Material::Air &&
			particleWorld->getParticle(row, col).wetnessMultiplier >= 0.9f)
		{
			// choose flower to grow, blue being the rarest, pink being the most common
			ParticleWorld::Material flowerToGrow = ParticleWorld::Material::RedFlower;
			bool growFlower = false;
			if (growFlowerChance < 10)
			{
				flowerToGrow = ParticleWorld::Material::BlueFlower;
				growFlower = true;
			}
			else if (growFlowerChance < 50)
			{
				flowerToGrow = ParticleWorld::Material::YellowFlower;
				growFlower = true;
			}
			else if (growFlowerChance < 75)
			{
				flowerToGrow = ParticleWorld::Material::PinkFlower;
				growFlower = true;
			}
			else if (growFlowerChance < 100)
			{
				growFlower = true;
			}

			// grow the flower
			if (growFlower)
			{
				particleWorld->particles[get1DIndex(row - 1, col, particleWorld->colSize)].material = flowerToGrow;
			}
		}
	}

	if (particleWorld->canLeft(col))
	{
		// increment wetness if a liquid is present (left)
		if (particleWorld->getParticle(row, col - 1).materialType == ParticleWorld::MaterialType::Liquid &&
			particleWorld->getParticle(row, col).wetnessMultiplier <= 0.5f)
		{
			particleWorld->particles[get1DIndex(row, col, particleWorld->colSize)].wetnessMultiplier += 0.05f;
		}
	}

	if (particleWorld->canRight(col))
	{
		// increment wetness if a liquid is present (right)
		if (particleWorld->getParticle(row, col + 1).materialType == ParticleWorld::MaterialType::Liquid &&
			particleWorld->getParticle(row, col).wetnessMultiplier <= 0.5f)
		{
			particleWorld->particles[get1DIndex(row, col, particleWorld->colSize)].wetnessMultiplier += 0.05f;
		}
	}

	if (particleWorld->canDown(row))
	{
		// apply gravity (down)
		if (
			(particleWorld->getParticle(row + 1, col).materialType == ParticleWorld::MaterialType::Gas ||
				particleWorld->getParticle(row + 1, col).materialType == ParticleWorld::MaterialType::Liquid) &&
			particleWorld->getParticle(row, col).material == self.material)
		{
			particleWorld->setParticle(row + 1, col, self);
			particleWorld->resetParticle(row, col);
		}
	}

	if (particleWorld->canLeft(col) && particleWorld->canDown(row))
	{
		// apply gravity (left)
		if (
			(particleWorld->getParticle(row + 1, col - 1).materialType == ParticleWorld::MaterialType::Liquid ||
				particleWorld->getParticle(row + 1, col - 1).materialType == ParticleWorld::MaterialType::Gas) &&
			particleWorld->getParticle(row, col - 1).material == ParticleWorld::Material::Air &&
			particleWorld->getParticle(row, col).material == self.material)
		{

			particleWorld->setParticle(row + 1, col - 1, self);
			particleWorld->resetParticle(row, col);
		}
	}

	if (particleWorld->canRight(col) && particleWorld->canDown(row))
	{
		// apply gravity (right)
		if (
			(particleWorld->getParticle(row + 1, col + 1).materialType == ParticleWorld::MaterialType::Liquid ||
				particleWorld->getParticle(row + 1, col + 1).materialType == ParticleWorld::MaterialType::Gas) &&
			particleWorld->getParticle(row, col + 1).material == ParticleWorld::Material::Air &&
			particleWorld->getParticle(row, col).material == self.material)
		{

			particleWorld->setParticle(row + 1, col + 1, self);
			particleWorld->resetParticle(row, col);
		}
	}
}